sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -209,20 +479,291 @@ node add_malloc_device(void *&data, size_t numBytes, const std::vector<node>& de
+| Returns the state of the queue
+
+|===
+
+A default constructed event is returned when a user submits a command-group to
+a queue in the recording state. These events have status
+`info::event_command_status::complete` and a user waiting on them will return
+immediately.
+
+==== Queue Properties
+
+:queue-properties: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:queue-properties
+
+There are {queue-properties}[two properties] defined by the core SYCL
+specification that can be passed to a `sycl::queue` on construction via the
+property list parameter. They interact with this extension in the following
+ways:
+
+1. `property::queue::in_order` - When a queue is created with the in-order
+   property, recording its operations results in a straight-line graph, as each
+   operation has an implicit dependency on the previous operation. However,
+   a graph submitted to an in-order queue will keep its existing structure such
+   that the complete graph executes in-order with respect to the other
+   command-groups submitted to the queue.
+
+2. `property::queue::enable_profiling` - This property has no effect on graph
+   recording. When set on the queue a graph is submitted to however, it allows
+   profiling information to be obtained from the event returned by a graph
+   submission.
+
+For any other queue property that is defined by an extension, it is the
+responsibility of the extension to define the relationship between that queue
+property and this graph extension.
+
+==== New Queue Member Functions
+
+Table 8. Additional member functions of the `sycl::queue` class.
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source, c++]
+----
+using namespace ext::oneapi::experimental;
+bool queue::begin_recording(command_graph<graph_state::modifiable> &graph)
+----
+
+|Synchronously changes the state of the queue to the `queue_state::recording`
+state.
+
+Parameters:
+
+* `graph` - Graph object to start recording commands to.
+
+Returns: `true` if the queue was previously in the `queue_state::executing`
+state, `false` otherwise.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the queue is already
+  recording to a different graph.
+
+|
+[source, c++]
+----
+bool queue::end_recording()
+----
+
+|Synchronously changes the state of the queue to the `queue_state::executing`
+state.
+
+Returns: `true` if the queue was previously in the `queue_state::recording`
+state, `false` otherwise.
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+event queue::submit(command_graph<graph_state::executable> graph)
+----
+
+|When invoked with the queue in the `queue_state::recording` state, a graph is
+added as a subgraph node. When invoked with the queue in the default
+`queue_state::executing` state, the graph is submitted for execution. Support
+for submitting a graph for execution, before a previous execution has been
+completed is backend specific. The runtime may throw an error.
```

Comment:
"Support for submitting a graph for execution, before a previous execution has been
completed is backend specific. The runtime may throw an error."

I used the original SYCL_EXT_ONEAPI_GRAPH wording here, but SYCL_EXT_CODEPLAY_GRAPHS says

"There are no gurantees that more than one instance of `graph` will execute concurrently.
 Submitting a graph for execution, before a previous execution has been completed
 may result in serialized execution depending on the SYCL backend and
 characteristics of the graph."

We should discuss what we want to do here.

Reply:
Now tracked in https://github.com/reble/llvm/issues/9